ftw/news/portlets/news_portlet.py

```diff
@@ -0,0 +1,354 @@
+from Acquisition import aq_parent, aq_inner
+from DateTime import DateTime
+from ftw.news import _
+from ftw.news import utils
+from plone.app.portlets.interfaces import IPortletPermissionChecker
+from plone.app.portlets.portlets import base
+from plone.directives import form
+from plone.directives.form.form import SchemaAddForm, SchemaEditForm
+from plone.formwidget.autocomplete.widget import AutocompleteMultiFieldWidget
+from plone.formwidget.contenttree import MultiContentTreeFieldWidget
+from plone.formwidget.contenttree import PathSourceBinder
+from plone.portlets.interfaces import IPortletDataProvider
+from Products.CMFCore.utils import getToolByName
+from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile
+from z3c.form import button
+from z3c.form import form as z3cform
+from z3c.relationfield import RelationChoice
+from zope import schema
+from zope.component import getMultiAdapter
+from zope.interface import implements, invariant, Invalid
+
+
+class INewsPortletSchema(form.Schema):
+    portlet_title = schema.TextLine(
+        title=_(u'news_portlet_title_label', default=u'Title'),
+        description=u'',
+        required=True,
+        default=u'')
+
+    form.widget(filter_by_path=MultiContentTreeFieldWidget)
+    filter_by_path = schema.List(
+        title=_(u'news_portlet_filter_path_label', default=u'Limit to path'),
+        description=_(u'news_portlet_filter_path_description',
+                      default=u'Only show news items from a specific path.'),
+        value_type=RelationChoice(
+            source=PathSourceBinder(
+                navigation_tree_query={'is_folderish': True},
+                is_folderish=True
+            ),
+        ),
+        required=False,
+        missing_value=(),
+    )
+
+    current_context = schema.Bool(
+        title=_(u'news_portlet_filter_current_context_label',
+                default=u'Limit to current context'),
+        description=_(u'news_portlet_filter_current_context_description',
+                      default=u'Only show news items from the current '
+                              u'context.'),
+        default=True,
```

Comment:
This is actually a bug: when ``current_context`` is ``True``, the path of the currently rendered context is used.
When a context is accessed, which has a default page, the default page's path is used in the query.
When the default page is not folderish (which is usually the case), it cannot contain any news and therefore the query has no results.
This happens on the site root in a Plone default environment.

Using the canonical object fixes this issue.